### PR TITLE
back bid/ask support based on whale price

### PR DESCRIPTION
**Pull request by DonDuala**
Enter a second set of gamma orders a tick better than where there are size bid/offers on an illiquid book. If someone sweeps prices on an aggregator, then the risk manager gets filled first at this "off-market" price

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -140,7 +140,10 @@ export const MANGO_DOWNTIME_THRESHOLD = 15; // Time in Minutes to stop routing t
 export const fundingThreshold = 0.25; // % Funding to switch to openbook orders
 export const maxMktSpread = 0.01; // Max % allowable spread to use midvalue as accurate price
 export const gammaCompleteThreshold = 0.90; // Move on to new gamma cycle when 90% filled
-export const searchSteps = 7; // Number of times to reduce the jupiter qty to check for routes
+export const searchSteps = 5; // Number of times to reduce the jupiter qty to check for routes
```
Comment (brittcyr) [inline src/config.ts:143]: in the future, split off changes like this into a separate change

File: src/config.ts [hunk 2]
```diff
@@ -140,7 +140,10 @@ export const MANGO_DOWNTIME_THRESHOLD = 15; // Time in Minutes to stop routing t
...
 export const minExecutionPremium = 0.001; // Min USDC amount to fill MM on
 export const JUPITER_LIQUIDITY = 100; // USDC amount to check jupiter mid point price
-export const jupiterSlippage = 50; // bps to allow calculating mid price to succeed
\ No newline at end of file
+export const jupiterSlippage = 50; // bps to allow calculating mid price to succeed
+export const dimQty = 10000; // qty to find in the order book to dim
```
Comment (brittcyr) [inline src/config.ts:147]: mention in the comment that units are dollars

File: src/utils.ts [hunk 3]
```diff
@@ -209,6 +209,9 @@ export async function getPythPrice(splMint: PublicKey) {
   for (const symbol of data.symbols) {
     const price = data.productPrice.get(symbol)!;
     if (tokenToPythSymbol(splMintToToken(splMint)) == symbol) {
+      if (price == undefined){
+        return;
```
Comment (brittcyr) [inline src/utils.ts:213]: why this?

File: src/scalper.ts [hunk 4]
```diff
@@ -925,14 +959,57 @@ export class Scalper {
...
     gammaOrders.add(gammaAskTx.transaction);
+    
+    // Calculate effective gamma subject to max at price
+    const whaleBidDiff = fairValue - whaleBidPrice;
+    const whaleAskDiff = whaleAskPrice - fairValue;
+    const backBidPrice = whaleBidDiff > fairValue - gammaBid ? 
+      Math.floor((whaleBidPrice + this.tickSize) * (1 / this.tickSize)) / (1 / this.tickSize) : undefined;
+    const backAskPrice = whaleAskDiff > gammaAsk - fairValue ? 
+      Math.floor((whaleAskPrice - this.tickSize) * (1 / this.tickSize)) / (1 / this.tickSize) : undefined;
+    const whaleBidGammaQty = Math.min(netGamma * (maxBackGammaMultiple - 1), dipTotalGamma * whaleBidDiff - netGamma);
```
Comment (brittcyr) [inline src/scalper.ts:970]: add comments to explain these calculations or make the code speak for itself

File: src/scalper_utils.ts [hunk 5]
```diff
@@ -286,7 +286,7 @@ export async function getFairValue(
   symbol: string,
   owner: Keypair
   ) {
-  let fairValue: number;
+  let fairValue = 0;
```
Comment (brittcyr) [inline src/scalper_utils.ts:289]: is 0 just the value for no oracle price?
should document that or switch to undefined

**Review by brittcyr — APPROVED** [review]

File: src/scalper.ts (hunk 4 above)
Comment (DonDuala) [reply src/scalper.ts:970]: L963

Comment (brittcyr) [reply src/scalper.ts:970]: That doesnt really explain what is going on everywhere. Like what is
(1 / this.tickSize)) / (1 / this.tickSize)

that shows up in a few places

File: src/utils.ts (hunk 3 above)
Comment (DonDuala) [reply src/utils.ts:213]: catching a case observed with mngo where price was returning undefined and breaking. Can't seem to reproduce now

File: src/scalper.ts (hunk 4 above)
Comment (DonDuala) [reply src/scalper.ts:970]: rounding off to the min tick size for order entry, will add

File: src/scalper_utils.ts (hunk 5 above)
Comment (DonDuala) [reply src/scalper_utils.ts:289]: Correct, we pass fairvalue 0 if none of the oracles return a price. This was solving another break that can't reproduce now that the oracles are behaving

**Review by brittcyr — APPROVED** [review]

**Pull request merged**